arrays/arrays.c

```diff
@@ -69,43 +75,81 @@ void resize_array(Array *arr) {
  *
  * Throw an error if the index is out of range.
  *****/
-char *arr_read(Array *arr, int index) {
-
+char *arr_read(Array *arr, int index)
+{
+  int errnum;
   // Throw an error if the index is greater than the current count
-
+  if (arr->count > index)
+  {
+    errnum = errno;
+    fprintf(stderr, "Value of errno: %d\n", errno);
+    perror("Error printed by perror");
+    fprintf(stderr, "Given index is higher than the size of the array");
+    return -1;
+  }
   // Otherwise, return the element at the given index
+  else
+  {
+    return arr->elements[index];
+  }
 }
 
-
 /*****
  * Insert an element to the array at the given index
  *****/
-void arr_insert(Array *arr, char *element, int index) {
-
+void arr_insert(Array *arr, char *element, int index)
+{
+  int errnum;
   // Throw an error if the index is greater than the current count
+  if (index > arr->count)
+  {
+    errnum = errno;
+    fprintf(stderr, "Value of errno: %d\n", errno);
+    perror("Error printed by perror");
+    fprintf(stderr, "Given index is higher than the size of the array");
+    return -1;
+  }
 
   // Resize the array if the number of elements is over capacity
+  if (arr->capacity == arr->count)
```

Comment:
Generally speaking this shouldn't cause an issue, but because of the public nature of these variables, in my mind I can imagine someone could set arr->count to a number greater than arr->capacity before calling this function. Which means this condition would never be hit and the resizing would not take place as needed to prevent an segmentation fault or similar issue with accessing non-allocated memory. In other words I think it'd be best to use `arr->count >= arr->capacity` instead of the strictly equal to condition.